### DISCUSSION TITLE
Include scripts and styles only when component used

### DISCUSSION THREAD

**Discussion by y-nk** [withastro/roadmap#664] [Proposal]
### Body

# Summary

Stop including `<script>` tags in a page build when a component is not printed in the page.

# Background

This is coming from https://github.com/withastro/astro/issues/8223 at the demand of @bluwy ; I personally believe this is a bug, but as I understood from my issue it's rather a known caveat which i would like to discuss removing.

# TL;DR

I simply believe that this should not be a reality:

`Child.astro`
```html
<script>
import foo from './super-big-script'
console.log({ foo })
</script>
```

`src/pages/index.astro`
```tsx
---
import Child from '~/components/Child.astro'
console.log({ Child }) // ← this will force-include the <script> above in the html 
---

<html>
  <body>I should be light</body>
</html>
```

Coming from Next/Nuxt, this feels not natural and I believe most would file this as a bug.

# Motivation

The motivation is to solve a problem I'm having which comes from a pretty common usecase. 

I'm building a static website from a 3rd party CMS (in my case Strapi) for which page structure is dynamic.

When dealing with such as structure, you must have a `getComponentByType()` method which helps to compute the content of the page. When doing so, at some point you load component based on their `type: string` definition. There are 2 options:

### Dynamically

We build a map of "block type to import path" and load the components on demand.

  ```ts
  const map: Record<string, string> = {
    comp1: './Comp1.astro',
    compN: './CompN.astro',
  }
  
  export function getComponentByType(type: string) { 
    return (await import(map[type])).default
  }
  ```

When doing so, it cuts a few features. Notably, the `<script>` tag of the component loaded cannot be bundled and is not added to the page (because of the dynamic loading)

### Statically

We can build a map of "block type to component" and simply return the component.

  ```ts
  import Comp1 from './Comp1.astro'
  import CompN from './CompN.astro'

  const map: Record<string, Component> = {
    comp1: Comp1,
    compN: CompN,
  }
  
  export function getComponentByType(type: string) { 
    return map[type]
  }
  ```

When doing so, since every component is statically listed, every component of the map is included in the bundling of the page **even when they are actually not used in the page**.

This is very unusual compared to any other component libraries/framework out there, even tho i'm very aware of the unique angle of Astro.

# Goals

- Convince that, in the context of `<script>` tags used in a page, we should move away from static analysis to runtime analysis.
- Discuss/plan how to do it
- Convince to let me champion the PR, if my skills are deemed sufficient

**Comment by lilnasy** [comment]
You may be able to use a middleware and html-parse the entire response and remove scripts that way.

You would lose the benefit of streaming, but since you are concerned with static sites, it's acceptable to buffer the entire response body.

**Reply by y-nk** [reply]
@lilnasy i've mentioned on discord that i could eventually do that but i was stopped by somebody mentioning that some scripts (ex: from `Comp1` & `CompN`) could eventually be bundled together in one js file to save requests.

Even if I could, without the ability to add data attributes to script tags, they wouldn't be easily spottable in the final page right?

(and i agree on the streaming loss, it's definitely acceptable)

**Reply by lilnasy** [reply]
In that case, this is what I have in mind for a proposal — letting users pick their compromise.

```astro
---
export const streaming = false
---
```

Once astro gets this signal that you don't consider streaming important, it can replace the static analysis heuristic of finding styles and scripts, for hooks into component render calls.

Within scope:
- No more extra css and js (your case)
- No more missing css and js (content collection users, see withastro/astro#7761)
- Maybe open the door to CSS-in-JS solutions.

If this makes sense, I could share my notes tomorrow on discord.

**Reply by y-nk** [reply]
that would be for pages with a `prerender` value resolved to `true` only? could we reuse that boolean instead to bring automatic optimization to those pages?

**Reply by lilnasy** [reply]
> That would be for pages with prerender

No, the two flags would be independent. I don't see anything that should prevent it.

> Could we reuse

No, that would be breaking. It has to be new and opt-in.

**Reply by y-nk** [reply]
understood.

my 1st reason for asking is that i was wondering how a streaming boolean could be understood by a user when they use an adapter which don't support streaming, i thought it could be misleading or deceptive (or rtfm maybe :D)?

i understand the non-breaking "opt-in" approach ; it'd be nice if it could be defined both granularly at file level and globally at config level (with priority on the most granular of course).

**Comment by matthewp** [comment]
Thanks for bringing this up! I think there's something *possible* here, but with some caveats.

1. Runtime analysis would mean, in my mind, building these scripts at runtime. This would require including Vite at runtime. I think this is out of the question for Astro, at the moment.
2. We could build the scripts at build time, but then only inject them when they are used. The downsides I can see are:
    a. Do we only inject them *once* or every time the component is used. I think the former is what you want, but this is tricky to do in Astro at the moment.
    b. This would mean that they are injected where the component is used, not in the head like today. So the scripts would be fetched later, slightly delaying the time until they are active.

**Reply by bluwy** [reply]
I think no2 is what we can aim for. The Astro compiler is able to track the hoisted scripts entrypoint during build-time, and during runtime we can just link the specifier to the built chunk, similar to how we handle Astro island renderers too.

> 2\. a. Do we only inject them _once_ or every time the component is used.

I could see both being useful, maybe an option to choose between hoisting or not, like a middle-ground between `is:inline` 🤔 And for hoisting, we could lend our head propagation handling to hoist it to the `<head>`.

**Reply by lilnasy** [reply]
What's the use case for injecting a hoisted script once per component use?

**Reply by matthewp** [reply]
I don't know of a use-case. Just noting it because attemptingto only do it once is tricky in Astro's codebase due to parallel rendering. This has caused bugs in the past when doing similar things (to only add the island scripts once, for example). But the codebase has improved since then so maybe it's not as much a problem any more.

**Reply by y-nk** [reply]
@matthewp i do see what you mean, and yes including vite at runtime does definitely not sound good. i have no further idea forward but i was wondering if any of you knew about how Vue approaches this in comparison. what i mean is, vue is running on vite, capable to SSG and having SFC (even tho the format differs i would imagine it is processed in a similar way). do they also have these issues in Nuxt, for example?

**Comment by adbl** [comment]
Adding some related observations to this. I'm not sure if this is the same cause, but I'm seeing this even for files included with `Astro.glob`, example:

- `index.astro`: Iterates over `.mdx` files using `Astro.glob`, rendering links to each `page.url`.
- If the `mdx` imports any components with client scripts, the scripts are included in `index.astro` html too.

That was surprising. I tried to use `optimizeHoistedScript: false` without any difference.

**Comment by lilnasy** [comment]
I think a neat solution is possible here.

```astro
---
import Comp1 from './Comp1.astro'
import CompN from './CompN.astro'
import * as Namespace from 'library'

const map: Record<string, Component> = {
  comp1: Comp1,
  compN: CompN,
}
const Comp = pickOne(map)
const Heading = random ? "h1" : "h2"

$$propagate($$result, Comp, Heading, Namespace.Group, Namespace.GroupItem)
---
<Heading/>
<Comp/>
<Namespace.Group>
    <Namespace.GroupItem/>
</Namespace.Group>
```

`$$propagate` would be an internal function the that users wont see. It would go right before the template and include as arguments all the identifiers the compiler notices as being used in it. Its behavior would be to accumulate scripts and styles required by the components.

It would need help from both the compiler and the rendering runtime, but Vite does not need to be involved, and it includes the scripts only once per page inside the `<head>` element.

I came up with the idea thinking of a way to combine statically added assets and propagated assets (content collections). Unfortunately, the idea doesn't quite achive that. We can wait for the `$$propagate` call in the frontmatter of the main page but not for imported astro components because doing that would compromise on time to first bite. So static analysis would stay and be used for  every component that's not a page.

**Reply by y-nk** [reply]
let me rephrase to see if i get it: `$$propagate` would be something like `(ctx: PageAssetContext, ...childrenComponents: AstroComponent[]) => void` and the idea would then be that `$$propagate` introspects each item of `childrenComponents: AstroComponent[]` and accumulate asset data in the `PageAssetContext`. then this data would be handled by the framework as of how to bundle it all.

**Reply by y-nk** [reply]
assuming i understood in the correct way, i think this `$$propagate` would be injected only when `export const streaming = false` not to disturb TTFB for SSR projects (linked to our prior discussion)?

**Reply by lilnasy** [reply]
> let me rephrase to see if i get it

Yes, that's exactly it! The one thing I would clarify is that bundling has already happened, `$$propagate` would act at render-time to pick and choose the already bundled assets.

> export const streaming = false

The current idea is similar to this earlier idea in implementation, but the difference is that user does not need to flag anything.

The opt-out streaming wouldve meant that we perform runtime checks everywhere and import checks nowhere.

Here, we perform the reliable but slow runtime check only on the main page because we already wait for it, so TTFB is not affected. And the fast and crude import checks are performed for all the imported components because we send out the headers and the head element before their execution completes.

**Comment by lilnasy** [comment]
Implemented as an integration
https://github.com/lilnasy/gratelets/tree/main/packages/dynamic-import

**Comment by forberg** [comment]
:wave: 

This might have come up before, but I'll give it a shot:
So I've been seeing some feedback about codesplitting CSS for "dynamic components". And I guess it's hard to do when moving stylesheets to the `<head>` of the page since Astro does a static analysis of the CSS to include.
But could Astro inline the stylesheets for various components as you do with `directRenderScripts`? Maybe not a global option, but per component?
`<Component client:load styles:inline={{ blocking: "rendering", fetchpriority: "low", as: "link" }}" />` (same goes for Astro-components, but without `client:load` ofc)
The output could be smth like:
```html
<astro-island>
<link rel="styleeshet" href="Component.bundle.hash.css" blocking="rendering" fetchpriority="low">
<div>...Rest of the component html</div>
</astro-island>
```

With response streaming, I think this could work pretty well. It seems you can mitigate FOUC in various browsers (like FF), ref Jake Archibalds blogpost from some years ago: https://jakearchibald.com/2016/link-in-body/ 

Ref `blocking="rendering"`: https://chromestatus.com/feature/5452774595624960

There might be some performance penalities in some cases, but at least it might be up to the developer to "fine grain" this for each project 😊 WDYT?

**Reply by bluwy** [reply]
This is something I investigated in the past, but the reason I didn't make it inline like `directRenderScripts` is because

1. It's more common to write scoped styles than scripts, so you'd get a ton of links or style tags.
2. CSS would be handled in two different ways (via imports or Astro's custom handling). So you'd have to understand how they work differently and it would be confusing for some.
3. And there's more risk to incorrect CSS ordering especially when combined with no2.

You mentioned making this opt-in so this would resolve no1, but technically I don't think that opt-in syntax could work because during compilation of the Astro file itself, it needs to decide beforehand how the CSS will be emitted and controlled. A different opt-in syntax would be putting it inside the Astro file itself as a compiler hint, and that could work, but the other concerns about still apply.

In most case, since these are scoped styles, they're harmless if they're accidentally included on a page unused. For scripts it's more dangerous since you may not intend to run that script on every page.

**Reply by forberg** [reply]
I see. But it's too bad that the main bundles of styles will grow as long as we add more components that should be dynamically imported.  
Maybe another alternative might be to look at how Loadable components do it with webpack:
https://loadable-components.com/

**Reply by bluwy** [reply]
I don't think the bundle size will grow this way, only that styles could be included in pages where it might be unused? As long as you're importing components normally where they're used, it shouldn't cause size regressions.
